### DISCUSSION TITLE
PSA: Importing old saves may be troublesome. (was: food management changes - experiments, regressions, final commits?!  <= 0.E-10395-g4525b4f )

### DISCUSSION THREAD

**Discussion by Anaglyphic** [CleverRaven/Cataclysm-DDA#48621] [General]
I get that people want to make changes to stuff to make it more "this" or "that" but please don't break shi... stuff doing so!

This is, dare I say it, RETARDED. 

_(yes I know that is a trigger word for sensitive people who worry their loved ones are being persecuted, discriminated against, but it's not. It's a pair word to advance/advanced. It's from chemistry. And I'm sorry people use it for an insult, but I'm taking the damn word back, because it's useful and it's not the sole realm of hatred, it's just describing something that is retarded, something behind what it_ should _be.)_

Even in up to as far as the latest Jenkins autobuild bot from 04/15, 0.E-10395-g4525b4f, both windows or linux, some of this food inventory stuff is so irritating I wonder what the experience of a new user is like. From my perspective even I've considered going back to 0.E-2 and let whoever keeps making these commits ease up on the pot smoking. :sunglasses: 

Scenario for a new world, fresh start, new player not used to such things from catadda yet... they won't be using h/l/, yet they'll be manually flagging each item with arrow keys or numpad and hitting enter:

Ack! Run! Running through woods to lose stragglers. Finally. I can see a few on the far side of this house, but if I'm quiet... very thirsty and very hungry will turn much worse soon. So sneak into the house silently as possible. Hit up that kitchen cabinet, SCORE! Tons of canned food. Try to pick some up...

... nope.

OK fine, micro-manage gone wild huh? Manually stash your zombie basher so your BOTH your hands are free; 95% of your carry volume and weight are empty and your STR is 10 so you are kind of confused but assume you can somehow (miraculously!) lift a medium can of tuna fish or beans.

... still nope.

You press g, you tag just one single medium tin can > beans, and what happens? Pop-up:

```
┌───────────────────────────────────────────────────────────────────────────────────┐
│  Can't stash medium tin can > beans while it's not empty                          │
├───────────────────────────────────────────────────────────────────────────────────┤
│  w  Wield medium tin can > beans                                                  │
│  s  Spill contents of medium tin can > beans, then pick up medium tin can > beans │
└───────────────────────────────────────────────────────────────────────────────────┘

```
Huh?

OK. Fine, let's try that. WIELD that can, you never know, it might make an OK projectile.
So while holding one can of food let's try to nab another...

... and still nope.

```
┌───────────────────────────────────────────────────────────────────────────────────────────────┐
│  Can't stash medium tin can > canned corn while it's not empty                                │
├───────────────────────────────────────────────────────────────────────────────────────────────┤
│  w  Dispose of medium tin can > beans and wield medium tin can > canned corn                  │
│  s  Spill contents of medium tin can > canned corn, then pick up medium tin can > canned corn │
└───────────────────────────────────────────────────────────────────────────────────────────────┘

```
ARGH. That makes no sense! 

You cannot un-can canned corn and carry it! That's not gameplay immersion, that's just stupid.

sidebar: A medium tin can, that has stats of 0.50 L an 1.28 lbs. Uhh... have you ever SEEN a 16oz/500ml medium tin can before? It's pretty hard to make them weigh 1.28 lbs WITH contents, even heavy dense foodstuffs like clam chowder or meat chili... but whatever. Let's just say the reason the medium tin can of frickin' beans weighs 4.41 lbs is because of gameplay not reality, because that is some BS. _(if it IS supposed to be realistic, whoever did the kg/lb conversion math really EFF'ed up.)_ 

Seriously.

We are too close to the fire barrel, but if a new player couldn't immediately google and find out WTH is going on? 

They're done.

Thoughts?

**Comment by robotic-gamer** [comment]
seems like a bug. the issue seems to be those cans aren't sealed for some reason. with them sealed you should be able to store them in your backpack like a rag. without being sealed it would be like trying to store a glass of water in your backpack. try unloading the cans, the food may be solid in which case you can store that in your backpack and the empty can if you want. or if its liquid try storing it in a resealable container like a bottle or jar. by the way what are you using to store stuff? your storage item must also be able to accommodate the length of object you want to store not just its volume.

**Reply by Anaglyphic** [reply]
Yep I get that, but it's a can. :smile_cat: And that's assuming the person isn't noticing the red text up at **INDV Vol x.xx/x.xx  Lng x.xx/x.xx** But yea it is a legitimate question, thanks for trying to help. Plenty of storage both volume and length (MOLLE and duffel mostly vacant) 

Yea I don't know, I'll just scrap the world and regen. Frustrating annoyance though.

The conversation was more about, for a reboot of a game pushing a decade, with still very few (relatively few) fans or players, it needs to make sure such things cannot happen. If the majority of the userbase was coding-capable, had github accounts, used linux 90% of their computing day, etc. knowing a 0.x release isn't a release... but it's not. You get on discourse or reddit and you'll see we are the minority it's windows then mac (non-technical) and then linux. Dozens, maybe a few hundred. Compared to the ~2500 Windows commenters. :man_dancing: :boom: :gun:

A first experience like that will send you back to a more GUI apocalypse like Metro: Exodus. lol

**Comment by Anaglyphic** [comment]
SHIT!

You're right it's an error, I found out what it is. It's apparently loading old saves from 0.E-2 or 0.E-3 not being imported correctly!

The crazy part is it does something to the world, to mapgen parameters.... it sometimes makes the error persistent! I remake the world but it'll persist. Making a new world, doesn't. 

I noticed this because I was having an issue with vehicles, you couldn't see through the windows if you were 1 tile away... back off, and you can see fine. Crazy. But before I came to ask WTH is going on, I noticed it was _all glass_, houses too...What a bizarre error cascade this causes. Wonder what else it breaks besides food containers and visual occlusion?

OK well, problem solved. Don't import old games, don't regen worlds whack them and make new.
Seems like a FAQ entry to make if someone ever gets around to making one. Wonder exactly what happened.

Thanks for letting me think out loud, as it were.

edit - removed all the images and stacked them to make a GIF for the subreddit to see if others had any idea what's up.

![cdda-when-is-glass-not-glass](https://user-images.githubusercontent.com/19985100/116496448-14e21b80-a873-11eb-9e20-c5262630f9f5.gif)

**Reply by actual-nh** [reply]
@Anaglyphic: Old saves (at least back to the original 0.E) are supposed to work; please put in an issue, or at least contribute the saves for use in testing. (See #42658.)

**Reply by Anaglyphic** [reply]
Thanks, I will. I'll make a more terse issue and link the two...

edit: https://github.com/CleverRaven/Cataclysm-DDA/issues/48694

When is glass not glass?
-----------------

Sorry this is long and filled with rambling. I didn't want to post a rambling thing that sounds like a request for tech support in open issues, so I'm putting this in a discussion thread and linking to it.

So yea, the picking up canned food thing is now secondary. And it hasn't happened since trying to load an old save, so it could just be mods enabled in one and not the newer or something.

But I have gotten the "glass thing" twice now, and makes driving impossible so is kind of big.

I guess 2x is repeatable, and it is persistent so it's part of the world save somewhere... just don't know exactly how to trigger it or where it's at.

Hell I can't even describe it well. "glass thing" being where glass isn't transparent unless you're >1 tile away. Imagine a window was not see-through if you were at the window, but if you step back you can see through it. I noticed it when I hopped into my fully restored Humvee and couldn't see out any of the windows. See dumb animated GIF above.

So, new world, new character, new install directory - no existing config or save, and extracted from latest Jenkins autobuild and played from that directory. (downloaded the "d" variant which I assume is debug...)

Everything starts normal, at least glass is acting as glass. I snuck into the nearest city and worked my way toward a gun store only to find no way in, reinforced with metal all around - pretty substantial Day 1 obstacle. However it was paired with another building, a church, where you could go upstairs, walk along the roof ledge and on to the adjoining building. Go down it's stairs, and you're in the (quite well-protected) gun store, and metal reinforcement remains undamaged by cutting torch.

After one of my multiple up/across/down, up/across/down scavenging excursions, after you feel safe and are not so precise about every single tile step, I finally noticed there was even more to the church? temple? whatever, a 3rd floor, a little rectory with a desk and couch and food and a bed. And because of both the two sets of stairs and how narrow the area, also makes for a good hidey hole. I don't know if the scentmap is 3D or they can smell me as if I were on the ground floor, but still easy to defend. 

I started sleeping up here, and some time afterwards - just a couple in-game days. I noticed I had "it" again. Note I don't know if they are connected, just pointing out when I noticed it. And even if it were, "causation is not correlation" is true in everything. It's also possible that since AFAIK the stained glass church windows are are above ground level, you can't see out when you're up close... need to be 1 tile away. Or like behind a big chair; obscured vision til you back off 1 tile. The exact thing I am experiencing, but  with all glass. 

So I came back to find anyone with a similar issue, but it's just me? How is that possible?

As for logging there's nothing much. debug.log seems more about crashes; lots of startup and INFO notifications and addresses, but in ~10 hours of gameplay only 2 logged errors - one about an item losing it's position and another about multiple zombies trying to enter the same tile and no one called _"FIRST!"_ afterward. Neither game-breaking and neither are likely related to this visual issue. I'll post them anyway because whey not?

```
09:55:51.079 ERROR : src/item_location.cpp:782 [void item_location::deserialize(JsonIn&)] parent location does not point to valid item
09:55:52.210 ERROR : src/item_location.cpp:290 [bool item_location::impl::item_on_person::ensure_who_unpacked() const] Failed to find item_location owner with character_id 1
09:55:52.727 ERROR : src/item_location.cpp:290 [bool item_location::impl::item_on_person::ensure_who_unpacked() const] Failed to find item_location owner with character_id 1
09:55:54.923 ERROR : src/item_location.cpp:114 [void item_location::impl::ensure_unpacked() const] item_location lost its target item during a save/load cycle
...
00:34:18.373 ERROR : src/creature_tracker.cpp:77 [bool Creature_tracker::add(shared_ptr_fast<monster>&)] there's already a monster at 12,128,0
00:34:18.568 ERROR : src/creature_tracker.cpp:77 [bool Creature_tracker::add(shared_ptr_fast<monster>&)] there's already a monster at 12,129,0

```
I have no idea what it is only what it's doing. Sorry for being so wordy, I didn't know if I was getting the concept across. 

Since it's happened 2x to me already, I think for now, before diving into source or setting up watched variables, I'll just start making snapshot backups while I'm playing; since the files are text we can compress the hell out of them; early even midgame entire worlds of close dense size 16 cities zips down to a megabyte or two? Something like:

**watch -n 600 'tar -cvjf 0_savesnapshots_0/TestWorld02_save_$(date +%m-%d-%y_%H-%M-%S.%3N).tar.bz2 -C TestWorld02/ .'**

Then I can play until I notice it happens again, and work backwards restoring world directory snapshots til I find the "working / broken" pair and at least have a better starting point.

Thanks for reading!

**Comment by actual-nh** [comment]
Can you search the log for "crouch", "walk", and "run"?

**Comment by kevingranade** [comment]
> (yes I know that is a trigger word for sensitive people who worry their loved ones are being persecuted, discriminated against, but it's not. It's a pair word to advance/advanced. It's from chemistry. And I'm sorry people use it for an insult, but I'm taking the damn word back, because it's useful and it's not the sole realm of hatred, it's just describing something that is retarded, something behind what it should be.)

No, fuck off with your abelist bullshit.
